ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java

```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+        }
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToDrop.getStartTimeInMs());
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
+        }
+      }
+      requestsToDrop.removeAll(errorRequests);
+    } while (requestToServe == null && requestsToDrop.isEmpty());
+
+    return new NetworkRequestBundle(requestToServe, requestsToDrop);
+  }
+
   @Override
-  public NetworkRequest receiveRequest() throws InterruptedException {
-
-    NetworkRequest request = null;
-    while (request == null) {
-      request = requestQueue.take();
-      http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - request.getStartTimeInMs());
-      if (request.equals(EmptyRequest.getInstance())) {
-        logger.debug("Request handler {} received shut down command ", request);
-      } else {
-        DataInputStream stream = new DataInputStream(request.getInputStream());
-        try {
-          // The first 8 bytes is size of the request. TCP implementation uses this size to allocate buffer. See {@link BoundedReceive}
-          // Here we just need to consume it.
-          stream.readLong();
-        } catch (IOException e) {
-          logger.error("Encountered an error while reading length out, close the connection", e);
-          closeConnection(request);
-          request = null;
+  public List<NetworkRequest> getUnqueuedRequests() throws InterruptedException {
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    NetworkRequest unqueuedRequest;
+    do {
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        requestsToDrop.add(unqueuedRequest);
+      }
+      if (requestsToDrop.isEmpty()) {
+        // If there are no un-queued requests, wait on the queue.
+        unqueuedRequest = unqueuedRequests.take();
+        requestsToDrop.add(unqueuedRequest);
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
         }
       }
+      requestsToDrop.removeAll(errorRequests);
+    }  while (requestsToDrop.isEmpty());
+    return requestsToDrop;
+  }
+
+  /**
+   * Consumes the first 8 bytes of the input stream in the incoming request since TCP implementation uses this size to
+   * allocated Buffer.
+   * @param request incoming network request.
+   * @return false if there was any error while reading the first 8 bytes. Else, return true.
+   */
+  private boolean consumeStream(NetworkRequest request) throws InterruptedException {
```

Comment:
this just seems like some verification method right ?

Reply:
better name would verifyXXXHeader ?

Reply:
These are actually to consume/discard the first 8 bytes. Added few java docs comments and renamed it to `consumeSizeHeader`. Hopefully it is less confusing now.